lib/teleterm/clusters/cluster_kubes.go

```diff
@@ -111,22 +114,56 @@ type GetKubesResponse struct {
 }
 
 // reissueKubeCert issue new certificates for kube cluster and saves them to disk.
-func (c *Cluster) reissueKubeCert(ctx context.Context, kubeCluster string) error {
+func (c *Cluster) reissueKubeCert(ctx context.Context, kubeCluster string, mfaPrompt mfa.Prompt) (tls.Certificate, error) {
 	// Refresh the certs to account for clusterClient.SiteName pointing at a leaf cluster.
 	err := c.clusterClient.ReissueUserCerts(ctx, client.CertCacheKeep, client.ReissueParams{
 		RouteToCluster: c.clusterClient.SiteName,
 		AccessRequests: c.status.ActiveRequests.AccessRequests,
 	})
 	if err != nil {
-		return trace.Wrap(err)
+		return tls.Certificate{}, trace.Wrap(err)
 	}
 
-	// Fetch the certs for the kube cluster.
-	return trace.Wrap(c.clusterClient.ReissueUserCerts(ctx, client.CertCacheKeep, client.ReissueParams{
-		RouteToCluster:    c.clusterClient.SiteName,
-		KubernetesCluster: kubeCluster,
-		AccessRequests:    c.status.ActiveRequests.AccessRequests,
-	}))
+	proxyClient, err := c.clusterClient.ConnectToProxy(ctx)
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+	defer proxyClient.Close()
+
+	var mfaRequired bool
+	key, err := proxyClient.IssueUserCertsWithMFA(
+		ctx, client.ReissueParams{
+			RouteToCluster:    c.clusterClient.SiteName,
+			KubernetesCluster: kubeCluster,
+			RequesterName:     proto.UserCertsRequest_TSH_KUBE_LOCAL_PROXY,
+		},
+		mfaPrompt,
+		client.WithMFARequired(&mfaRequired),
+	)
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+
+	// TODO(ravicious): Add the equivalent of checkIfCertsAreAllowedToAccessCluster.
+
+	// TODO(ravicious): Share this logic with issueKubeCert in tool/tsh/common/kube_proxy.go.
+	if !mfaRequired {
+		if err := c.clusterClient.LocalAgent().AddKubeKey(key); err != nil {
+			return tls.Certificate{}, trace.Wrap(err)
+		}
+	}
```

Comment:
Makes sense, I'm just going to remove it then. I was just copying stuff from tsh without thinking much. It turns out there's not much code to reuse between tsh and teleterm then. Well, excluding the mountains of code we already reuse in the form of local proxies and kube middleware.

Do you know why tsh does this? Is it just to avoid unnecessary cert issues if the cert is already there?

I suppose Connect could do the same as well, but as of today this is not implemented in any gateway.

Reply:
> Do you know why tsh does this? Is it just to avoid unnecessary cert issues if the cert is already there?

The certs are shared between all `tsh` kube commands. It makes `tsh proxy kube` load quicker but is not necessary IMO. For other commands like `kubectl` (`tsh kube credentials`) or `tsh kubectl`, it is more noticeable whether the certs are reloaded every time or not.

At least my experience with Connect, since there is a GUI where you click things, see tabs opening etc., it doesn't feel slow at all with the current implementation.